### PR TITLE
Fixes #38 - Compute a mobile TSCI

### LINKED ISSUE #38: Split out Desktop + Mobile bugs
We could treat desktop bugs as a range of columns, and the same for mobile, then graph them separately or as a composite thing.

I guess this will effectively 2x our report generation time.

### PR DESCRIPTION
**Pull request by miketaylr**
Right now this just filters the existing results into mobile-specific counterparts (and feels super un-DRY, but whatever...).

Next up: jam it into the spreadsheets.

### CODE REVIEW

**Comment by miketaylr** [comment]
Here's an example: https://docs.google.com/spreadsheets/d/1e4H82LOPXYW_fMECjajLBAfiA2lKXfZjfxPnfI3Mz4k/edit?usp=sharing

(note, I filed a bug for the last row, it comes from master #69)

**Comment by miketaylr** [comment]
> This branch has conflicts

lol

**Comment by past** [comment]
> > This branch has conflicts
> 
> lol

I won! 🏆

**Comment by past** [comment]
In the example run it seems weird that duplicates(mobile) > duplicates.

**Comment by miketaylr** [comment]
> In the example run it seems weird that duplicates(mobile) > duplicates.

Ah, interesting. I'll double check.

File: bugs.js [hunk 1]
```diff
@@ -254,9 +302,13 @@ const getWebcompat = async (website, githubKey, minDate, maxDate) => {
...
         // filter out any bugs with an sci-exclude label
         .filter(bug => bug.labels.every(label => label.name !== "sci-exclude"));
+    const filteredMobileResults = filteredResults.filter(isMobileWebCompat);
+    const filteredMobileCriticalResults = filteredCriticals.filter(isMobileWebCompat);
     return {
         webcompatResult: `=HYPERLINK("${webcompatQuery}"; ${filteredResults.length})`,
-        criticalsResult: `=HYPERLINK("${criticalsQuery}"; ${filteredCriticals.length})`
+        criticalsResult: `=HYPERLINK("${criticalsQuery}"; ${filteredCriticals.length})`,
+        webcompatMobileResult: `=HYPERLINK("${webcompatQuery}"; ${filteredMobileResults.length})`,
+        criticalsMobileResult: `=HYPERLINK("${criticalsQuery}"; ${filteredMobileCriticalResults.length})`,
```
Comment (past) [inline bugs.js]: Are we using the same link for both mobile and total because the guthub API doesn't support the necessary query parameters? I think you've said something to that effect in another bug, but I can't find it right now.

File: bugs.js [hunk 2]
```diff
@@ -376,29 +427,39 @@ const getDuplicates = async (website, bugzillaKey, githubKey, minDate, maxDate)
...
     for (const [ query, ghToBzMap ] of searches) {
         const milestoneSearch = `https://api.github.com/search/issues?q=${query}`;
         const results = await getAllGitHubResultsFor(octokit.request, {url: milestoneSearch});
         for (const item of results) {
             const bzId = ghToBzMap.get(item.number);
-            if (bzId && item.milestone.title === "duplicate") {
+            if (isMobileWebCompat(item) && bzId && item.milestone.title === "duplicate") {
+                dupedMobileBzIds.add(bzId);
+                dupedMobileGhIds.add(item.number);
+            } else if (bzId && item.milestone.title === "duplicate") {
```
Comment (past) [inline bugs.js]: I would have gone with the more DRY nested if clause, but this works, too.

**Comment by miketaylr** [comment]
OK, new test sheet: https://docs.google.com/spreadsheets/d/1eum0f3PdZJkSo477h9pTRa2hKpSbqm_t1sBr0aDxWG0/edit?usp=sharing

@past I'm gonna head out for the day, but I'll check back tonight in case this is merge able. No issue if not, but I can kick-off a run for new data to run throughout the weekend.

File: bugs.js [hunk 3]
```diff
@@ -376,29 +427,40 @@ const getDuplicates = async (website, bugzillaKey, githubKey, minDate, maxDate)
...
+    const dupedMobileGhIds = new Set();
     const dupedBzIds = new Set();
+    const dupedMobileBzIds = new Set();
     const octokit = getOctokitInstance(githubKey);
     for (const [ query, ghToBzMap ] of searches) {
         const milestoneSearch = `https://api.github.com/search/issues?q=${query}`;
         const results = await getAllGitHubResultsFor(octokit.request, {url: milestoneSearch});
         for (const item of results) {
             const bzId = ghToBzMap.get(item.number);
+            if (isMobileWebCompat(item) && bzId && item.milestone.title === "duplicate") {
```
Comment (past) [inline bugs.js]: I would have gone with the more DRY nested if clause, but this works, too.

File: bugs.js [hunk 4]
```diff
@@ -376,29 +427,40 @@ const getDuplicates = async (website, bugzillaKey, githubKey, minDate, maxDate)
...
     }
     const bzLink = `https://bugzilla.mozilla.org/buglist.cgi?o1=anyexact&v1=${param}&f1=bug_id`;
-    return `=HYPERLINK("${bzLink}"; ${dupedGhIds.size})`;
+    let mobileParam = "";
+    for (const id of dupedMobileBzIds) {
+        mobileParam += "%2C" + id;
+    }
+    const bzMobileLink = `https://bugzilla.mozilla.org/buglist.cgi?o1=anyexact&v1=${mobileParam}&f1=bug_id`;
+    return {
+        duplicatesResult: dupedGhIds.size ? `=HYPERLINK("${bzLink}"; ${dupedGhIds.size})`: 0,
```
Comment (past) [inline bugs.js]: So this is a slight behavior change, but I'm not 100% sure if this was intentional. Previously we were checking for `dupedBzIds.size` being zero, but now we are testing `dupedGhIds.size` instead. Apparently the only difference is that the latter may be smaller if the milestone query filters out some github issues that weren't in the duplicate milestone. In that case, we should indeed use the more strict set for the result.

Is the above accurate?

File: index.js [hunk 5]
```diff
@@ -67,10 +67,10 @@ const main = async () => {
     for (const date of queryDates) {
         const LIST_FILE = await tranco.fetchList(LIST_SIZE, LIST_DIR, date);
         const bugTable = await bugs.fetchBugs(LIST_FILE, bugzillaKey, githubKey, parsedMinDate, date);
-        if (bugTable.get("bugzilla").length +
-            bugTable.get("webcompat").length +
-            bugTable.get("criticals").length +
-            bugTable.get("duplicates").length === 0) {
+        if (bugTable.get("bugzilla").length + bugTable.get("bugzillaMobile").length +
```
Comment (past) [inline index.js]: I don't think this change is necessary. This check is for the weird edge case where sometimes the tranco list will be empty. In that case all the main (total) results will be zero. Since by definition the mobile results are subsets of the total results, they will be zero, too. Therefore we only need to look at the total results to see if we are hitting this edge case.

Come to think of it, we could even simplify this check by only considering bugzilla and webcompat results only. If these two are zero, there is no way that there will be criticals and duplicates with non-zero values.

File: spreadsheet.js [hunk 6]
```diff
@@ -154,8 +202,10 @@ async function addStaticData(sheets, spreadsheetId, listSize, listFile = 'data/l
         "startColumnIndex": 0,
         "endColumnIndex": 2
     };
-    const headers = ['Rank', 'Website', 'bugzilla', 'webcompat.com',
-        'criticals', 'duplicates', 'critical weight', 'SCI', 'Site weight', 'Weighted SCI'];
+    const headers = ['Rank', 'Website', 'bugzilla', 'bugzilla (mobile)', 'webcompat.com',
+        'webcompat.com (mobile)', 'criticals', 'criticals (mobile)', 'duplicates',
+        'duplicates (mobile)', 'critical weight', 'SCI', 'SCI (mobile)', 'Site weight',
+        'Weighted SCI', 'Weighted SCI (mobile)'];
```
Comment (past) [inline spreadsheet.js]: Optional, but how would you feel about shortening the titles a bit so that they can fit in the cell? Instead of, say "webcompat.com (mobile)", perhaps something like "webcompat.com (mob)" or "webcompat.com (M)", or even "webcompat-M".

File: bugs.js [hunk 7]
```diff
@@ -38,11 +46,27 @@ const fetchBugs = async (listFile = 'data/list.csv', bugzillaKey, githubKey, min
...
-        duplicates.push(await getDuplicates(website, bugzillaKey, githubKey, minDate, maxDate));
-        const { webcompatResult, criticalsResult } = await getWebcompat(website, githubKey, minDate, maxDate);
+        const {
+            bugzillaResult,
+            bugzillaMobileResult
+        } = await getBugzilla(website, bugzillaKey, minDate, maxDate);
+        bugzilla.push(bugzillaResult);
+        bugzillaMobile.push(bugzillaMobileResult);
+        const {
+            duplicatesResult, duplicatesMobileResult
```
Comment (wisniewskit) [inline bugs.js]: Might as well split this onto two lines if we're going to format this way ;)

File: spreadsheet.js [hunk 8]
```diff
@@ -74,17 +98,41 @@ async function addBugData(sheets, spreadsheetId, bugTable, title) {
...
         resource: {
             values: [duplicates],
             majorDimension: "COLUMNS",
         },
         valueInputOption,
     })
     console.log('Updated duplicates cells: ' + result.data.updatedCells);
+
+    const duplicatesMobile = bugTable.get("duplicatesMobile");
+    result = await sheets.spreadsheets.values.update({
```
Comment (wisniewskit) [inline spreadsheet.js]: Is it worth factoring out this repeated boilerplate into a simpler function like `result = await updateSheetRange(spreadsheetId, title, range, values, valueInputOption)`?

File: bugs.js (hunk 1 above)
Comment (miketaylr) [reply bugs.js]: Yeah, precisely. It's not great, but I'm not sure how we can get around it -- currently you can't build up a query using labels, and have it function as OR. The last one wins, so the results are misleading.

File: bugs.js (hunk 2 above)
Comment (miketaylr) [reply bugs.js]: oh yeah, we can do that.

File: bugs.js (hunk 4 above)
Comment (miketaylr) [reply bugs.js]: Oh, I think this is just a mistake on my end -- accepting VSCode's suggestion without paying too much attention. Will change.

File: index.js (hunk 5 above)
Comment (miketaylr) [reply index.js]: SGTM, will change.

File: spreadsheet.js (hunk 6 above)
Comment (miketaylr) [reply spreadsheet.js]: I had this thought too, I was going to go for something silly like `webcompat.com 📱) -- but that emoji is kinda hard to parse). I'll go for something short like (M) or dash-M.

File: spreadsheet.js (hunk 8 above)
Comment (miketaylr) [reply spreadsheet.js]: Probably!

File: bugs.js (hunk 4 above)
Comment (past) [reply bugs.js]: If my explanation above is accurate however, it would make your change an improvement :-)
I think @wisniewskit made the original check here, any thoughts?

File: bugs.js (hunk 3 above)
Comment (miketaylr) [reply bugs.js]: done.

File: bugs.js (hunk 4 above)
Comment (miketaylr) [reply bugs.js]: Oh right, now I see what you're saying. @wisniewskit, WDYT?

Comment (wisniewskit) [reply bugs.js]: Yes, it would be best to keep the test as Mike has it in this patch, for the reasons you gave, @past. (Basically I was only doing this to make sure no hyperlink was emitted for the zero-case, so it's best to match the variable it checks with the one it outputs).

File: spreadsheet.js (hunk 8 above)
Comment (miketaylr) [reply spreadsheet.js]: I'm going to file a follow-up refactor bug here, because I'm lazy.

Comment (miketaylr) [reply spreadsheet.js]: https://github.com/past/tsci/issues/71

File: bugs.js (hunk 4 above)
Comment (miketaylr) [reply bugs.js]: Thanks, let's all pretend I did this on purpose and we'll keep it as is.
